pkg/ebpf/dns_parser.go

```diff
@@ -0,0 +1,109 @@
+// +build linux_bpf
+
+package ebpf
+
+import (
+	"bytes"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+)
+
+const maxIPBufferSize = 200
+
+type dnsParser struct {
+	decoder *gopacket.DecodingLayerParser
+	layers  []gopacket.LayerType
+	payload *layers.DNS
+
+	// Cached translation object to reduce allocations
+	cached *translation
+}
+
+func newDNSParser() *dnsParser {
+	payload := &layers.DNS{}
+
+	stack := []gopacket.DecodingLayer{
+		&layers.Ethernet{},
+		&layers.IPv4{},
+		&layers.IPv6{},
+		&layers.UDP{},
+		payload,
+	}
+
+	return &dnsParser{
+		decoder: gopacket.NewDecodingLayerParser(layers.LayerTypeEthernet, stack...),
+		payload: payload,
+		cached:  new(translation),
+	}
+}
+
+func (p *dnsParser) Parse(data []byte) *translation {
+	err := p.decoder.DecodeLayers(data, &p.layers)
+	if err != nil || p.decoder.Truncated {
+		return nil
+	}
+
+	for _, layer := range p.layers {
+		if layer == layers.LayerTypeDNS {
+			return p.parseAnswer(p.payload)
+		}
+	}
+
+	return nil
+}
+
+// source: https://github.com/weaveworks/scope
+func (p *dnsParser) parseAnswer(dns *layers.DNS) *translation {
+	// Only consider responses to singleton, A-record questions
+	if !dns.QR || dns.ResponseCode != 0 || len(dns.Questions) != 1 {
+		return nil
+	}
+	question := dns.Questions[0]
+	if question.Type != layers.DNSTypeA || question.Class != layers.DNSClassIN {
+		return nil
+	}
+
+	var (
+		domainQueried = question.Name
+		records       = append(dns.Answers, dns.Additionals...)
+		translation   = p.getCachedTranslation(domainQueried)
+		alias         []byte
+	)
+
+	// Retrieve the CNAME record, if available.
+	for _, record := range records {
+		if record.Type == layers.DNSTypeCNAME && record.Class == layers.DNSClassIN &&
+			bytes.Equal(domainQueried, record.Name) {
+			alias = record.CNAME
+			break
+		}
+	}
+
+	// Get IPs
+	for _, record := range records {
+		if record.Type != layers.DNSTypeA || record.Class != layers.DNSClassIN {
+			continue
+		}
+		if bytes.Equal(domainQueried, record.Name) ||
+			(alias != nil && bytes.Equal(alias, record.Name)) {
+			translation.add(util.AddressFromNetIP(record.IP))
+		}
+	}
+
+	return translation
+}
+
+func (p *dnsParser) getCachedTranslation(dns []byte) *translation {
```

Comment:
Is it possible that there could be concurrency issues with this cached `*translation` object? 

As far as I see it, it will be returned via `Parse` and stored in the `SocketFilterSnooper. reverseDNSCache`.  

However, it looks like _alll_ parse requests will also use that same cached object which will result in previously returned translations objects being modified on new parse requests.  Does that sound right ?

Reply:
So all parsing and cache inserts happen within the same go-routine. That's the trade-off we have to do in order to use zero-copy, otherwise we would have to copy the data somehow.
As for what is being referenced inside the cache, the two things allocated in the heap from the `dnsParser` are:
* the DNS string
* each individual util.Address
but none of those are modified. 

The `*translation` struct and the underlying `[]util.Address` are the things that get re-used and modified, but as I said, there is no concurrent access on that.

Reply:
Do you need to run https://golang.org/pkg/runtime/#LockOSThread on this goroutine or is the same goroutine fine enough?

Reply:
Good point. AFAIK there is any thread locality concerns involved with polling or referencing that the ring buffer memory, but if that was the case I would expect it to be referenced in the `afpacket` documentation, which is *not* the case.